File: app/src/main/java/ch/unstable/ost/HeadNavigationFragment.java
package ch.unstable.ost;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.MainThread;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.TextView;

import java.util.Date;

import ch.unstable.ost.api.model.ConnectionQuery;
import ch.unstable.ost.utils.LocalizationUtils;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.base.Verify.verifyNotNull;


public class HeadNavigationFragment extends BaseNavigationFragment {


    private static final int REQUEST_CODE_CHOOSE_TO = 1;
    private static final int REQUEST_CODE_CHOOSE_FROM = 2;
    private static final String KEY_STATE = "KEY_STATE";
    private static final String TAG = "HeadNavigationFragment";
    private View.OnClickListener mOnButtonClickListener;
    private SelectionState mSelectionState;
    private Button mToButton;
    private Button mFromButton;
    private ImageButton mReverseDirectionButton;
    private TextView mTime;

    public HeadNavigationFragment() {
        // Required empty public constructor
    }


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);
        mOnButtonClickListener = new OnNavigationButtonsClickListener();

        mSelectionState = null;
        if (savedInstanceState != null) {
            mSelectionState = savedInstanceState.getParcelable(KEY_STATE);
        } else if (getArguments() != null) {
            mSelectionState = getArguments().getParcelable(KEY_STATE);
        }
        if (mSelectionState == null) {
            mSelectionState = new SelectionState();
        }
        Disposable disposable = mSelectionState.getChangeObservable()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(getSelectionStateObserver());
    }

    @NonNull
    private Consumer<SelectionState> getSelectionStateObserver() {
        return new Consumer<SelectionState>() {
            @Override
            public void accept(SelectionState selectionState) throws Exception {
                updateViews();
                onQueryChanged();
            }
        };
    }

    private void updateViews() {
        final Context context = getContext();
        if(context == null) return;
        mTime.setText(LocalizationUtils.getArrivalOrDepartureText(context,
                mSelectionState.getArrivalTime(),
                mSelectionState.getDepartureTime()));
        mToButton.setText(getToButtonText(context, mSelectionState.getTo()));
        mFromButton.setText(getFromButtonText(context, mSelectionState.getFrom()));
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        if (mSelectionState != null) {
            outState.putParcelable(KEY_STATE, mSelectionState);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_head_navigation, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mFromButton = view.findViewById(R.id.fromButton);
        mToButton = view.findViewById(R.id.toButton);
        mReverseDirectionButton = view.findViewById(R.id.reverseDirectionButton);
        mTime = view.findViewById(R.id.timeView);
        View timeSettingsContainer = view.findViewById(R.id.timeSettingsContainer);
        timeSettingsContainer.setOnClickListener(mOnButtonClickListener);
        mFromButton.setOnClickListener(mOnButtonClickListener);
        mToButton.setOnClickListener(mOnButtonClickListener);
        mReverseDirectionButton.setOnClickListener(mOnButtonClickListener);
        updateViews();
    }

    private void startStationChooser(@StringRes int chooseRequest, int codeTo) {
        String chooseString = getContext().getString(chooseRequest);
        Intent intent = new Intent(getContext(), ChooseStationActivity.class);
        intent.setAction(Intent.ACTION_RUN);
        intent.putExtra(ChooseStationActivity.EXTRA_CHOOSE_PROMPT, chooseString);
        startActivityForResult(intent, codeTo);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK) {
            String name;
            switch (requestCode) {
                case REQUEST_CODE_CHOOSE_FROM:
                    name = data.getStringExtra(ChooseStationActivity.EXTRA_RESULT_STATION_NAME);
                    //noinspection ResultOfMethodCallIgnored
                    verifyNotNull(name, "Choose station result (from) is null");
                    mSelectionState.setFrom(name);
                    break;
                case REQUEST_CODE_CHOOSE_TO:
                    name = data.getStringExtra(ChooseStationActivity.EXTRA_RESULT_STATION_NAME);
                    //noinspection ResultOfMethodCallIgnored
                    verifyNotNull(name, "Choose station result (to) is null");
                    mSelectionState.setTo(name);
                    break;
                default:
                    // Unknown result
            }
        }
    }

    @NonNull
    private static String getToButtonText(Context context, @Nullable String to) {
        if (to == null) {
            return context.getString(R.string.request_choose_to);
        }
        return to;
    }

    @NonNull
    private static String getFromButtonText(Context context, @Nullable String from) {
        if (from == null) {
            return context.getString(R.string.request_choose_from);
        }
        return from;
    }

    private void onQueryChanged() {
        if (mSelectionState.getTo() != null && mSelectionState.getFrom() != null) {
            ConnectionQuery query = SelectionState.createConnectionQuery(mSelectionState);
            selectRoute(query);
        }
    }

    private void onReverseDirectionRequested() {
        final Context context = getContext();
        Animation animation = AnimationUtils.loadAnimation(context, R.anim.half_rotation);
        mReverseDirectionButton.startAnimation(animation);
        final Animation fadeInTop = AnimationUtils.loadAnimation(context, R.anim.fade_in_top);
        final Animation fadeOutTop = AnimationUtils.loadAnimation(context, R.anim.fade_out_top);
        final Animation fadeInBottom = AnimationUtils.loadAnimation(context, R.anim.fade_in_bottom);
        final Animation fadeOutBottom = AnimationUtils.loadAnimation(context, R.anim.fade_out_bottom);
        final String newTo = mSelectionState.getFrom();
        final String newFrom = mSelectionState.getTo();
        fadeOutTop.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation1) {
                // Not interested in this event
            }

            @Override
            public void onAnimationEnd(Animation animation1) {
                mFromButton.setText(getFromButtonText(context, newFrom));
                mFromButton.startAnimation(fadeInTop);
                mToButton.setText(getToButtonText(context, newTo));
                mToButton.startAnimation(fadeInBottom);
                mSelectionState.setFrom(newFrom);
                mSelectionState.setTo(newTo);
            }

            @Override
            public void onAnimationRepeat(Animation animation1) {
                // Not interested in this event
            }
        });
        mFromButton.startAnimation(fadeOutTop);
        mToButton.startAnimation(fadeOutBottom);
    }

    @MainThread
    public void updateQuery(ConnectionQuery query) {
        mSelectionState.setQuery(query);
    }

    public void clearQuery() {
        mSelectionState.setFrom(null);
        mSelectionState.setTo(null);
    }

    public class OnNavigationButtonsClickListener implements View.OnClickListener {
        @Override
        public void onClick(View v) {
            switch (v.getId()) {
                case R.id.toButton:
                    startStationChooser(R.string.request_choose_to, REQUEST_CODE_CHOOSE_TO);
                    break;
                case R.id.fromButton:
                    startStationChooser(R.string.request_choose_from, REQUEST_CODE_CHOOSE_FROM);
                    break;
                case R.id.reverseDirectionButton:
                    onReverseDirectionRequested();
                    break;
                case R.id.timeSettingsContainer:
                    onOpenTimeSettings();
                    break;
            }
        }

        private void onOpenTimeSettings() {

            Date date = mSelectionState.getDepartureTime();
            TimePickerDialog.TimeRestrictionType restrictionType = TimePickerDialog.TimeRestrictionType.DEPARTURE;
            if (date == null && mSelectionState.getArrivalTime() != null) {
                restrictionType = TimePickerDialog.TimeRestrictionType.ARRIVAL;
                date = mSelectionState.getArrivalTime();
            }

            TimePickerDialog timePickerDialog = new TimePickerDialog(getContext(), restrictionType, date, new TimePickerDialog.OnTimeSelected() {
                @Override
                public void onArrivalTimeSelected(@NonNull Date date) {
                    mSelectionState.setArrivalTime(date);
                }

                @Override
                public void onDepartureTimeSelected(@NonNull Date date) {
                    mSelectionState.setDepartureTime(date);
                }
            });
            timePickerDialog.show();
        }
    }
}

File: app/src/main/java/ch/unstable/ost/HeadNavigationFragment.kt
package ch.unstable.ost

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import androidx.annotation.MainThread
import androidx.annotation.StringRes
import ch.unstable.ost.TimePickerDialog.OnTimeSelected
import ch.unstable.ost.TimePickerDialog.TimeRestrictionType
import ch.unstable.ost.api.model.ConnectionQuery
import ch.unstable.ost.databinding.FragmentHeadNavigationBinding
import ch.unstable.ost.utils.LocalizationUtils.getArrivalOrDepartureText
import com.google.common.base.Verify
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Consumer
import java.util.*

class HeadNavigationFragment : BaseNavigationFragment() {
    private var onButtonClickListener: View.OnClickListener? = null
    private var selectionState: SelectionState? = null
    private var binding: FragmentHeadNavigationBinding? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        onButtonClickListener = OnNavigationButtonsClickListener()
        selectionState = null
        if (savedInstanceState != null) {
            selectionState = savedInstanceState.getParcelable(KEY_STATE)
        } else if (arguments != null) {
            selectionState = requireArguments().getParcelable(KEY_STATE)
        }
        if (selectionState == null) {
            selectionState = SelectionState()
        }
        val disposable = selectionState!!.getChangeObservable()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(selectionStateObserver)
    }

    private val selectionStateObserver: Consumer<SelectionState>
        private get() = Consumer {
            updateViews()
            onQueryChanged()
        }

    private fun updateViews() {
        val context = context ?: return
        val binding = binding ?: return
        binding.timeView.text = getArrivalOrDepartureText(
                context,
                selectionState!!.getArrivalTime(),
                selectionState!!.getDepartureTime())
        binding.toButton.text = getToButtonText(context, selectionState!!.getTo())
        binding.fromButton.text = getFromButtonText(context, selectionState!!.getFrom())
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        if (selectionState != null) {
            outState.putParcelable(KEY_STATE, selectionState)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val binding = FragmentHeadNavigationBinding.inflate(inflater, container, false)
        this.binding = binding
        binding.timeSettingsContainer.setOnClickListener(onButtonClickListener)
        binding.fromButton.setOnClickListener(onButtonClickListener)
        binding.toButton.setOnClickListener(onButtonClickListener)
        binding.reverseDirectionButton.setOnClickListener(onButtonClickListener)
        updateViews()
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        this.binding = null
    }

    private fun startStationChooser(@StringRes chooseRequest: Int, codeTo: Int) {
        val chooseString = requireContext().getString(chooseRequest)
        val intent = Intent(context, ChooseStationActivity::class.java)
        intent.action = Intent.ACTION_RUN
        intent.putExtra(ChooseStationActivity.EXTRA_CHOOSE_PROMPT, chooseString)
        startActivityForResult(intent, codeTo)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == Activity.RESULT_OK) {
            val name: String
            when (requestCode) {
                REQUEST_CODE_CHOOSE_FROM -> {
                    name = data!!.getStringExtra(ChooseStationActivity.EXTRA_RESULT_STATION_NAME)
                    Verify.verifyNotNull(name, "Choose station result (from) is null")
                    selectionState!!.setFrom(name)
                }
                REQUEST_CODE_CHOOSE_TO -> {
                    name = data!!.getStringExtra(ChooseStationActivity.EXTRA_RESULT_STATION_NAME)
                    Verify.verifyNotNull(name, "Choose station result (to) is null")
                    selectionState!!.setTo(name)
                }
                else -> {
                }
            }
        }
    }

    private fun onQueryChanged() {
        if (selectionState!!.getTo() != null && selectionState!!.getFrom() != null) {
            val query = selectionState!!.createQuery()
            selectRoute(query)
        }
    }

    private fun onReverseDirectionRequested() {
        val context = context
        val animation = AnimationUtils.loadAnimation(context, R.anim.half_rotation)
        binding?.reverseDirectionButton?.startAnimation(animation)
        val fadeInTop = AnimationUtils.loadAnimation(context, R.anim.fade_in_top)
        val fadeOutTop = AnimationUtils.loadAnimation(context, R.anim.fade_out_top)
        val fadeInBottom = AnimationUtils.loadAnimation(context, R.anim.fade_in_bottom)
        val fadeOutBottom = AnimationUtils.loadAnimation(context, R.anim.fade_out_bottom)
        val newTo = selectionState!!.getFrom()
        val newFrom = selectionState!!.getTo()
        fadeOutTop.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation1: Animation) {
                // Not interested in this event
            }

            override fun onAnimationEnd(animation1: Animation) {
                val binding = binding
                if(binding != null) {
                    binding.fromButton.text = getFromButtonText(context, newFrom)
                    binding.fromButton.startAnimation(fadeInTop)
                    binding.toButton.text = getToButtonText(context, newTo)
                    binding.toButton.startAnimation(fadeInBottom)
                }
                selectionState!!.setFrom(newFrom)
                selectionState!!.setTo(newTo)
            }

            override fun onAnimationRepeat(animation1: Animation) {
                // Not interested in this event
            }
        })
        binding?.fromButton?.startAnimation(fadeOutTop)
        binding?.toButton?.startAnimation(fadeOutBottom)
    }

    @MainThread
    fun updateQuery(query: ConnectionQuery?) {
        selectionState!!.setQuery(query!!)
    }

    fun clearQuery() {
        selectionState!!.setFrom(null)
        selectionState!!.setTo(null)
    }

    inner class OnNavigationButtonsClickListener : View.OnClickListener {
        override fun onClick(v: View) {
            when (v.id) {
                R.id.toButton -> startStationChooser(R.string.request_choose_to, REQUEST_CODE_CHOOSE_TO)
                R.id.fromButton -> startStationChooser(R.string.request_choose_from, REQUEST_CODE_CHOOSE_FROM)
                R.id.reverseDirectionButton -> onReverseDirectionRequested()
                R.id.timeSettingsContainer -> onOpenTimeSettings()
            }
        }

        private fun onOpenTimeSettings() {
            var date = selectionState!!.getDepartureTime()
            var restrictionType = TimeRestrictionType.DEPARTURE
            if (date == null && selectionState!!.getArrivalTime() != null) {
                restrictionType = TimeRestrictionType.ARRIVAL
                date = selectionState!!.getArrivalTime()
            }
            val timePickerDialog = TimePickerDialog(context!!, restrictionType, date, object : OnTimeSelected {
                override fun onArrivalTimeSelected(date: Date) {
                    selectionState!!.setArrivalTime(date)
                }

                override fun onDepartureTimeSelected(date: Date) {
                    selectionState!!.setDepartureTime(date)
                }
            })
            timePickerDialog.show()
        }
    }

    companion object {
        private const val REQUEST_CODE_CHOOSE_TO = 1
        private const val REQUEST_CODE_CHOOSE_FROM = 2
        private const val KEY_STATE = "KEY_STATE"
        private const val TAG = "HeadNavigationFragment"
        private fun getToButtonText(context: Context?, to: String?): String {
            return to ?: context!!.getString(R.string.request_choose_to)
        }

        private fun getFromButtonText(context: Context?, from: String?): String {
            return from ?: context!!.getString(R.string.request_choose_from)
        }
    }
}